File: downloader/src/player.py
import math


class Player:

    def __init__(self, name: str):
        self.rating = 1500
        self.name = name

    def update_rating(self, change: int, score: int, team_scores: [int]):
        gamma = 6
        team_score = sum(team_scores)
        score_ratio = score / team_score
        if change > 0:
            sa = 3 * (math.pow(score_ratio, gamma) /
                      (math.pow(team_scores[0] / team_score, gamma) +
                       math.pow(team_scores[1] / team_score, gamma) +
                       math.pow(team_scores[2] / team_score, gamma)))
        else:
            sa = 3 * (1 / math.pow(score_ratio, gamma) /
                      (1 / (math.pow(team_scores[0] / team_score, gamma)) +
                       (1 / math.pow(team_scores[1] / team_score, gamma)) +
                       (1 / math.pow(team_scores[2] / team_score, gamma))))
        self.rating += change * sa

File: src/player.rs

#[derive(Debug, Clone)]
pub struct Player {
    pub name: String,
    pub rating: i32,
    pub games_played: i32,
}
impl Player {
    pub fn new(name: String) -> Player {
        Player {
            name,
            rating: 1500,
            games_played: 0,
        }
    }

    pub fn update_rating(&mut self, change: i32, score: i32, team_scores: Vec<i32>) {
        let gamma = 6;
        let team_score = team_scores.iter().sum::<i32>();
        let score_ratio = score as f32 / team_score as f32;

        if change > 0 {
            let sa = 3.0 * (score_ratio.powf(gamma as f32) as f32 /
                ((team_scores[0] as f32 / team_score as f32).powf(gamma as f32) as f32 +
                (team_scores[1] as f32 / team_score as f32).powf(gamma as f32) as f32 +
                (team_scores[2] as f32 / team_score as f32).powf(gamma as f32) as f32));
            self.rating += (change as f32 * sa) as i32;
        }
        else {
            let sa = 3.0 * ((1.0 / score_ratio).powf(gamma as f32) as f32 /
                ((1.0 / (team_scores[0] as f32 / team_score as f32)).powf(gamma as f32) as f32 +
                (1.0 / (team_scores[1] as f32 / team_score as f32)).powf(gamma as f32) as f32 +
                (1.0 / (team_scores[2] as f32 / team_score as f32)).powf(gamma as f32) as f32));
            self.rating += (change.abs() as f32 * sa) as i32;
        }
        self.games_played += 1;
    }
}